### PR TITLE
Add libjxl (jpegxl) library components

### PR DESCRIPTION
**Pull request by hmaarrfk**
Continuation of: https://github.com/conda-forge/staged-recipes/pull/17812

Bisecting : MACOSX_DEPLOYMENT_TARGET

- 12.3 good
- 11.0 good
- 10.12 good
- 10.9 fail

Bisecting: MACOSX_SDK_VERSION
  - 12.3 good
  - 11.0 good
  - 10.12 good
  - 10.9 unknown

Other cleanups

- [ ] minimize patch

Checklist
- [ ] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [ ] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [ ] Source is from official source.
- [ ] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [ ] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [ ] Build number is 0.
- [ ] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [ ] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by hmaarrfk** [comment]
Maybe... linux needs something like:
```cmake
pkg_check_modules(Imath IMPORTED_TARGET Imath)                                                                                                                                                                        
if (Imath_FOUND)                                                                                                                                                                                                      
  target_include_directories(jxl_extras_core-obj PRIVATE                                                                                                                                                              
    "${Imath_INCLUDE_DIRS}"                                                                                                                                                                                           
  )                                                                                                                                                                                                                   
  target_compile_definitions(jxl_extras_core-obj PRIVATE -DJPEGXL_ENABLE_EXR=1)                                                                                                                                       
  list(APPEND JXL_EXTRAS_CODEC_INTERNAL_LIBRARIES PkgConfig::Imath)                                                                                                                                                   
endif() 
```

but that didn't work locally, so i'm not sure.

File: recipes/jpegxl/bld.bat [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    -DJPEGXL_ENABLE_EXAMPLES=OFF   ^
+    -DJPEGXL_BUNDLE_GFLAGS=OFF     ^
+    -DJPEGXL_ENABLE_JNI=OFF        ^
+    -DJPEGXL_ENABLE_OPENEXR=OFF    ^
+    -DJPEGXL_ENABLE_SKCMS=OFF      ^
+    ..
+
+if %ERRORLEVEL% neq 0 (type CMakeError.log && exit 1)
+
+cmake --build . --target install --config Release
```
Comment (hmaarrfk) [inline recipes/jpegxl/bld.bat:28]:
```suggestion
cmake --build . --target install --config Release  -- -j 1
```

just to validate that it isn't a multi-threading issue

File: recipes/jpegxl/conda_build_config.yaml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+MACOSX_SDK_VERSION:            # [osx and x86_64]
+  - "12.3"                     # [osx and x86_64]
+MACOSX_DEPLOYMENT_TARGET:      # [osx and x86_64]
+  - "12.3"                     # [osx and x86_64]
```
Comment (hmaarrfk) [inline recipes/jpegxl/conda_build_config.yaml:4]: need to bisect down to 10.9 10.12 to ensure that we need to bump so high.

File: recipes/jpegxl/bld.bat [hunk 3]
```diff
@@ -0,0 +1,29 @@
+@echo ON
+
+mkdir build
+cd build
+cmake -LAH -G "Ninja"              ^
+    -DCMAKE_BUILD_TYPE=Release     ^
```
Comment (hmaarrfk) [inline recipes/jpegxl/bld.bat:6]:
```suggestion
    -DJPEGXL_ENABLE_TOOLS=OFF ^
    -DCMAKE_BUILD_TYPE=Release     ^
```

File: recipes/jpegxl/build.sh [hunk 4]
```diff
@@ -0,0 +1,30 @@
+set -ex
+CXXFLAGS="-D__STDC_FORMAT_MACROS ${CXXFLAGS}"
+
+mkdir build
+cd build
+cmake \
+    -DCMAKE_BUILD_TYPE=Release     \
+    -DBUILD_TESTING=OFF            \
```
Comment (hmaarrfk) [inline recipes/jpegxl/build.sh:8]:
```suggestion
    -DJPEGXL_ENABLE_TOOLS=OFF      \
    -DBUILD_TESTING=OFF            \
```

**Comment by hmaarrfk** [comment]
@cgohlke, i'm kindly requesting your expertise in building this library.

Do you build out the tools for windows and osx or  do you explictely disable them with the flag 

```
JPEGXL_ENABLE_TOOLS=OFF
```

i'm wondering if they are useful at all to developers trying to compress with jpegxl for the first time or if you think it is OK to omit them? 

I could optionally try to build them for linux only, but it is quite frustrating to get them working on windows and OSX.

File: recipes/libjxl/conda_build_config.yaml [hunk 5]
```diff
@@ -0,0 +1,4 @@
+MACOSX_SDK_VERSION:            # [osx and x86_64]
+  - "12.3"                     # [osx and x86_64]
```
Comment (hmaarrfk) [inline recipes/libjxl/conda_build_config.yaml:2]:
```suggestion
  - "11.0"                     # [osx and x86_64]
```

**Comment by cgohlke** [comment]
> Do you build out the tools for windows and osx or do you explictely disable them with the flag

On Windows, I build the library and tools with the following CMake options (PNG is also picked up):
```
    -DBUILD_SHARED_LIBS=OFF ^
    -DBUILD_TESTING=OFF ^
    -DJPEGXL_STATIC=ON ^
    -DJPEGXL_ENABLE_DOXYGEN=OFF ^
    -DJPEGXL_ENABLE_TOOLS=ON ^
    -DJPEGXL_ENABLE_BENCHMARK=OFF ^
    -DJPEGXL_ENABLE_FUZZERS=OFF ^
    -DJPEGXL_ENABLE_EXAMPLES=OFF ^
    -DJPEGXL_ENABLE_MANPAGES=OFF ^
    -DJPEGXL_ENABLE_JNI=OFF ^
    -DJPEGXL_ENABLE_SKCMS=OFF ^
    -DJPEGXL_FORCE_SYSTEM_HWY=OFF ^
    -DJPEGXL_FORCE_SYSTEM_BROTLI=ON ^
    -DJPEGXL_FORCE_SYSTEM_LCMS2=ON ^
    -DJPEGXL_ENABLE_OPENEXR=ON
```
On other platforms, when building imagecodecs in Docker or CI, I disable the tools:
```
        -DCMAKE_BUILD_TYPE=Release \
        -DCMAKE_INSTALL_PREFIX="${build_dir}" \
        -DBUILD_SHARED_LIBS=ON \
        -DBUILD_TESTING=OFF \
        -DJPEGXL_ENABLE_BENCHMARK=OFF \
        -DJPEGXL_ENABLE_FUZZERS=OFF \
        -DJPEGXL_ENABLE_EXAMPLES=OFF \
        -DJPEGXL_ENABLE_JNI=OFF \
        -DJPEGXL_ENABLE_TOOLS=OFF \
        -DJPEGXL_ENABLE_DOXYGEN=OFF \
        -DJPEGXL_ENABLE_MANPAGES=OFF \
        -DJPEGXL_ENABLE_OPENEXR=OFF \
```

On macOS I had to use recent versions: `macos-13` with `MACOSX_DEPLOYMENT_TARGET: "10.14"`.

> i'm wondering if they are useful at all to developers trying to compress with jpegxl for the first time or if you think it is OK to omit them?

Yes, the tools are very useful.

**Comment by hmaarrfk** [comment]
> Yes, the tools are very useful.

Thanks for the feedback. I was hoping that wasn’t the case. 

I might try to split them off for a second package maybe. The most powerful tool for us is imagecodecs and tifffile and those seem to work with what I have here. 

The tools from jpegxl could wait for a follow up in the feedstock itself. 

I’m somewhat excited by the latest release 0.10 since apparently they resolved challenges with larger images!!!

**Comment by hmaarrfk** [comment]
Alright. I’m going to merge this and hopefully bring us closer to feature parity between this and pypi. I feel bad that you had to already address a few questions regarding the missing library.

**Comment by hmaarrfk** [comment]
Sigh. All this work already existed:

https://github.com/conda-forge/libjxl-split-feedstock

**Pull request closed without merging**